url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/cmdlib/BeanLibrary.java

```diff
@@ -0,0 +1,560 @@
+
+        try
+        {
+            // Try to interprete the property as
+            // an integer.
+            lIdx = Integer.parseInt(lPropName);
+        }
+        catch (NumberFormatException e1)
+        {
+            // Ignore this exception!
+        }
+
+        if(aTarget instanceof List)
+        {
+            final List lTargetList = (List) aTarget;
+            if(lIdx < 0)
+                throw new CommandException("Modifying list should use an index.");
+            if(lIdx >= lTargetList.size())
+                throw new CommandException("Array index out of range.");
+
+            // TODO: catch errors on this one.
+            lTargetList.set(lIdx, aVal);
+
+        }
+        else if(aTarget != null)
+        {
+            if(aTarget.getClass().isArray())
+            {
+                // Array access.
+                if(lIdx < 0)
+                    throw new CommandException("Modifying array should use an index.");
+                if(lIdx >= Array.getLength(aTarget))
+                    throw new CommandException("Array index out of range.");
+
+                // TODO: catch errors on this one.
+                Array.set(aTarget, lIdx, aVal);
+            }
+            else
+            {
+                try
+                {
+                    // JavaBean access.
+                    BeanInfo lInfo = Introspector.getBeanInfo(aTarget.getClass());
+                    PropertyDescriptor lProps[] = lInfo.getPropertyDescriptors();
+                    for (PropertyDescriptor lProp : lProps)
+                    {
+                        if(lProp.getDisplayName().equals(lPropName))
+                        {
+                            Method lMeth = lProp.getWriteMethod();
+                            if(lMeth == null)
+                                throw new CommandException("Property cannot be written.");
+
+                            if(lProp.getPropertyType() != aVal.getClass() && aVal instanceof String)
+                            {
+                                PropertyEditor lEditor = lProp.createPropertyEditor(aTarget);
+                                lEditor.setAsText((String) aVal);
+                            }
+                            else lMeth.invoke(aTarget, new Object[]{aVal});
+                        }
+                    }
+                }
+                catch (Exception e)
+                {
+                    throw new CommandException(e.getMessage(), e);
+                }
+            }
+        }
+    }
+
+    private void render(Object aObj, PrintWriter aWriter)
+            throws CommandException
+    {
+        if(aWriter == null) return;
+
+        if(aObj instanceof IContext)
+        {
+            IContext lCtx = (IContext) aObj;
+            Map<String, Object> lDump = lCtx.dumpBindings();
+            for(String lKey : lDump.keySet())
+            {
+                Object lVal = lDump.get(lKey);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.